providers/apis/defi/osmosis/client.go

```diff
@@ -0,0 +1,237 @@
+
+	var spotPriceResponse SpotPriceResponse
+	if err := json.NewDecoder(resp.Body).Decode(&spotPriceResponse); err != nil {
+		return SpotPriceResponse{}, err
+	}
+
+	c.apiMetrics.AddHTTPStatusCode(c.api.Name, resp)
+	return spotPriceResponse, nil
+}
+
+// MultiClientImpl is an Osmosis client that wraps a set of multiple Clients.
+type MultiClientImpl struct {
+	logger     *zap.Logger
+	api        config.APIConfig
+	apiMetrics metrics.APIMetrics
+
+	clients []Client
+}
+
+// NewMultiClient creates a new Client.
+func NewMultiClient(
+	logger *zap.Logger,
+	api config.APIConfig,
+	apiMetrics metrics.APIMetrics,
+	clients []Client,
+) (Client, error) {
+	if err := api.ValidateBasic(); err != nil {
+		return nil, fmt.Errorf("failed to validate config: %w", err)
+	}
+
+	if api.Name != Name {
+		return nil, fmt.Errorf("invalid config: name (%s) expected (%s)", api.Name, Name)
+	}
+
+	if !api.Enabled {
+		return nil, fmt.Errorf("invalid config: disabled (%v)", api.Enabled)
+	}
+
+	if apiMetrics == nil {
+		return nil, fmt.Errorf("invalid config: apiMetrics is nil")
+	}
+
+	return &MultiClientImpl{
+		logger:     logger,
+		api:        api,
+		apiMetrics: apiMetrics,
+		clients:    clients,
+	}, nil
+}
+
+// NewMultiClientFromEndpoints creates a new Client from a list of endpoints.
+func NewMultiClientFromEndpoints(
+	logger *zap.Logger,
+	api config.APIConfig,
+	apiMetrics metrics.APIMetrics,
+) (Client, error) {
+	if err := api.ValidateBasic(); err != nil {
+		return nil, fmt.Errorf("failed to validate config: %w", err)
+	}
+
+	if api.Name != Name {
+		return nil, fmt.Errorf("invalid config: name (%s) expected (%s)", api.Name, Name)
+	}
+
+	if !api.Enabled {
+		return nil, fmt.Errorf("invalid config: disabled (%v)", api.Enabled)
+	}
+
+	if apiMetrics == nil {
+		return nil, fmt.Errorf("invalid config: apiMetrics is nil")
+	}
+
+	clients := make([]Client, 0, len(api.Endpoints))
+	for _, endpoint := range api.Endpoints {
+		c, err := NewClient(api, apiMetrics, endpoint)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create client: %w", err)
+		}
+
+		clients = append(clients, c)
+	}
+
+	return &MultiClientImpl{
+		logger:     logger,
+		api:        api,
+		apiMetrics: apiMetrics,
+		clients:    clients,
+	}, nil
+}
+
+// SpotPrice delegates the request to all underlying clients and applies a filter to the
+// set of responses.
+func (mc *MultiClientImpl) SpotPrice(ctx context.Context, poolID uint64, baseAsset, quoteAsset string) (SpotPriceResponse, error) {
+	resps := make([]SpotPriceResponse, len(mc.clients))
+
+	var wg sync.WaitGroup
+	wg.Add(len(mc.clients))
+
+	for i, client := range mc.clients {
+		url := mc.api.Endpoints[i].URL
+
+		index := i
+		go func(index int, client Client) {
+			// Observe the latency of the request.
+			start := time.Now()
+			defer func() {
+				wg.Done()
+				mc.apiMetrics.ObserveProviderResponseLatency(mc.api.Name, metrics.RedactedEndpointURL(index), time.Since(start))
```

Comment:
is this metric going to be duplicated since the individual clients themselves already record this?